skimage/morphology/max_tree.py

```diff
@@ -0,0 +1,656 @@
+        This image can be of any type.
+    connectivity: unsigned int, optional
+        The neighborhood connectivity. The integer represents the maximum
+        number of orthogonal steps to reach a neighbor. It is 1 for
+        4-connectivity and 2 for 8-connectivity. Default value is 2.
+
+    Returns
+    -------
+    parent: ndarray, int64
+        The value of each pixel is the index of its parent in the ravelled
+        array
+
+    tree_traverser: 1D array, int64
+        The ordered pixel indices (referring to the ravelled array). The pixels
+        are ordered such that every pixel is preceded by its parent (except for
+        the root which has no parent).
+
+    References
+    ----------
+    .. [1] Salembier, P., Oliveras, A., & Garrido, L. (1998). Antiextensive
+           Connected Operators for Image and Sequence Processing.
+           IEEE Transactions on Image Processing, 7(4), 555-570.
+    .. [2] Berger, C., Geraud, T., Levillain, R., Widynski, N., Baillard, A.,
+           Bertin, E. (2007). Effective Component Tree Computation with
+           Application to Pattern Recognition in Astronomical Imaging.
+           In International Conference on Image Processing (ICIP) (pp. 41-44).
+    .. [3] Najman, L., & Couprie, M. (2006). Building the component tree in
+           quasi-linear time. IEEE Transactions on Image Processing, 15(11),
+           3531-3539.
+    .. [4] Carlinet, E., & Geraud, T. (2014). A Comparative Review of
+           Component Tree Computation Algorithms. IEEE Transactions on Image
+           Processing, 23(9), 3885-3895.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skimage.max_tree import build_max_tree
+
+    We create a small sample image (Figure 1 from [4]) and build the max-tree.
+
+    >>> image = np.array([[15, 13, 16], [12, 12, 10], [16, 12, 14]])
+    >>> P, S = build_max_tree(image, connectivity=2)
+
+    """
+    # User defined masks are not allowed, as there might be more than one
+    # connected component in the mask (and therefore not a single tree that
+    # represents the image). Mask here is an image that is 0 on the border
+    # and 1 everywhere else.
+    mask_shrink = np.ones([x - 2 for x in image.shape], bool)
+    mask = np.pad(mask_shrink, 1, mode='constant')
+
+    neighbors, offset = _validate_connectivity(image.ndim, connectivity,
+                                               offset=None)
+
+    # initialization of the parent image
+    parent = np.zeros(image.shape, dtype=np.int64)
+
+    # flat_neighborhood contains a list of offsets allowing one to find the
+    # neighbors in the ravelled image.
+    flat_neighborhood = _compute_neighbors(image, neighbors,
+                                           offset).astype(np.int32)
+
+    # pixels need to be sorted according to their grey level.
+    tree_traverser = np.argsort(image.ravel(),
+                                kind='quicksort').astype(np.int64)
```

Comment:
Why did you specify quicksort here?

Reply:
I have forgotten (but I think there was a reason they mentioned in one of the papers) ... would you suggest removing this? I must say, that I would rather prefer not to test the effect if different sorting algorithms now.

Reply:
@ThomasWalter my suggestion would be to leave the default sorting algorithm *unless* the tests fail when you remove this line. I suspect the paper mentioned some specific algorithm they use but the traversal should certainly not depend on which sorting algorithm is used!

Reply:
OK.